src/ir/stack-utils.h

```diff
@@ -0,0 +1,121 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef wasm_ir_stack_h
+#define wasm_ir_stack_h
+
+#include "ir/properties.h"
+#include "wasm-type.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace StackUtils {
+
+// Iterate through `block` and remove nops.
+void compact(Block* block);
+
+} // namespace StackUtils
+
+// Stack signatures are like regular function signatures, but they are used to
+// represent the stack parameters and results of arbitrary sequences of stacky
+// instructions. They have to record whether they cover an unreachable
+// instruction because their composition takes into account the polymorphic
+// results of unreachable instructions.
+struct StackSignature {
+  Type params;
+  Type results;
+  bool unreachable;
+
+  StackSignature()
+    : params(Type::none), results(Type::none), unreachable(false) {}
+  StackSignature(Type params, Type results, bool unreachable = false)
+    : params(params), results(results), unreachable(unreachable) {}
+
+  StackSignature(const StackSignature&) = default;
+  StackSignature& operator=(const StackSignature&) = default;
+
+  // Get the stack signature of `expr`
+  explicit StackSignature(Expression* expr);
+
+  // Get the stack signature of the range of instructions [first, last). The
+  // sequence of instructions is assumed to be valid, i.e. their signatures
+  // compose.
+  template<class InputIt>
+  explicit StackSignature(InputIt first, InputIt last)
+    : params(Type::none), results(Type::none), unreachable(false) {
+    // TODO: It would be more efficient to build the params in reverse order
+    while (first != last) {
+      *this += StackSignature(*first++);
+    }
+  }
+
+  // Return `true` iff `next` composes after this stack signature.
+  bool composes(const StackSignature& next) const;
+
+  // Whether a block with this stack signature could be typed with `sig`.
+  bool satisfies(Signature sig) const;
+
+  // Compose stack signatures. Assumes they actually compose.
+  StackSignature& operator+=(const StackSignature& next);
+  StackSignature operator+(const StackSignature& next) const;
+
+  bool operator==(const StackSignature& other) const {
+    return params == other.params && results == other.results &&
+           unreachable == other.unreachable;
+  }
+};
+
+// Calculates stack machine data flow, associating the sources and destinations
+// of all values in a block.
+struct StackFlow {
+  // The input (output) location at which a value of type `type` is consumed
+  // (produced), representing the `index`th input into (output from) instruction
+  // `expr`. `unreachable` is true iff the corresponding value is consumed
+  // (produced) by the polymorphic behavior of an unreachable instruction
+  // without being used by that instruction.
```

Comment:
Even with DCE, I think we would be able to encounter a case that looks like this:

```
(block (result f32 i64)
  (i32.const 0)
  (block [unreachable]
    (unreachable)
  )
)
```

In this example, we will want to be able to infer an `(i32) -> (f32, f64)` type for the `unreachable` inner block. To do that, we need to calculate the data flow into and out of the polymorphic inner block.

We could also prevent that situation by having more constraints and requiring more optimizations to be run before entering Stacky IR, that would make Stacky IR much less robust and discovering all the necessary constraints in the first place would be difficult. Ultimately we want Stacky IR to represent and facilitate reasoning about raw WebAssembly as closely as possible, and that includes reasoning about polymorphic unreachables.

Reply:
I assume that code fragment is poppy IR?

I don't think dce-d binaryen IR can lead to cases like that, and as we'll only get poppy IR from binaryen IR, I think that means we are ok to ignore unnecessary unreachable code (and I think unnecessary unreachable code is what's relevant here). But I may be missing something at several levels here...

Reply:
Looking forward to using wasp to parse WebAssembly directly to Poppy IR and unstackifying it in a separate step, we will need to be able to represent unreachable code in Poppy IR because it can be presesent in WebAssembly.

Overall I am uncomfortable with making the assumption that Poppy IR has already been DCEd. That just seems brittle to me because then all Poppy IR optimizations have to avoid accidentally introducing new unreachable code. It also limits Poppy IR's ability to faithfully represent any WebAssembly module.

Reply:
If we start out with the Binaryen IR

```
(block (result f32 i64)
 (i32.add
  (... side effects 1 ...) [i32]
  (block [unreachable]
   (unreachable)
   (.... side effects 2 ...)
  )
 )
)
```

Then DCE should be able simplify this to something like

```
(block [unreachable]
 (... side effects 1 ...) [i32]
 (block [unreachable]
  (unreachable)
 )
)
```
Which is similar to the fragment from the prior post. Type inference still needs to be able to figure out what concrete type to assign the unreachable blocks here, which requires being able to calculate the outflow from a polymorphic unreachable instruction.

Reply:
I believe DCE turns that into
```
(block (result f32 i64)
 (drop
  (... side effects 1 ...) [i32]
 )
 (unreachable)
)
```
(the first thing with side effects is dropped, and the unreachable block - which is unreachable code itself - is gone too). I don't think we need to infer any types in this case?

`wasm => poppy => binaryen IR` and the brittleness arguments are definitely valid arguments, though! My hope was that we can simplify poppy IR by a lot if it doesn't need unreachability at all. But we should consider those two arguments to see if they don't justify it.

Reply:
Aha, no wonder the DCE pass is so complicated. Maybe we can simplify it by making it use Poppy IR ;)

Actually, since these utilities give us the ability to do type inference for unreachable blocks, we might be able to simplify normal Binaryen IR's unreachability rules to match Poppy IR's rather than the other way around. Specifically, this would allow us to make any block unreachable if it is not branched to and it contains an unreachable instruction. This would get rid of the restriction that the final expression of the block cannot have a concrete type. That would be a follow-up investigation, of course, but it would make unreachability more powerful as well as simplify and unify the unreachability rules.

Reply:
Interesting idea about unreachability! Needs more thought but I don't see an obvious reason not to remove that restriction, if it helps.

I do think we should think about the other two arguments before moving forward here, though. Specifically, we should have an idea if we want to use poppy IR during read and not just during write (the latter is all I had in mind), and whether poppy passes will need to introduce new unreachable code (I was actually hoping they'd be very simple and not do that - they would focus on stack-machine specific stuff, all done after dead code was already removed).

Reply:
Places we've considered (with any amount of seriousness) using Poppy IR:

1. Pre-write optimizations: We can impose arbitrary structural restrictions (i.e. require arbitrary optimizations to already have run), but doing so makes testing, fuzzing, and general reasoning slightly more complicated.

2. Text parsing: Parsing to Poppy IR (probably via wasp) would allow us to support the full standard text format for the first time and to test the interpreter on the upstream spec tests. We would need to be able to represent arbitrary WebAssembly, including unreachable code. The unstackifier could be slightly simpler if stack utilities could reason correctly about this unreachable code, although the same logic could also be built into the unstackifier itself.

3. Binary parsing: We could outsource binary parsing to wasp as well, with the same considerations as text parsing.

4. Simplifying existing passes: It would be good to take a look at simplifying DCE, MergeBlocks, and any pass that expects flat IR by having them operate on Poppy IR, either having them run stackification and unstackification as subpasses or depending on the pass runner to insert conversions. Clearly we can't require running DCE before we run DCE, but more generally it is more robust to limit dependencies between optimization passes.

5. Unknown future passes: There's not much to say here except that the fewer limitations we put on Poppy IR, the more useful it can be.

Also note that Stack IR has no restrictions on unreachable code, so placing such restrictions on Poppy IR would make them more difficult to compare and would be a regression in capabilities.

I also don't think there is much upside to disallowing unreachable code in Poppy IR, since it behaves identically to unreachable instructions in the underlying WebAssembly that people should probably be familiar with anyways. This seems like a clear choice to me.

Reply:
> I also don't think there is much upside to disallowing unreachable code in Poppy IR

I guess this is the bottom line. If there isn't much cost to support it now, then I'm not opposed.

So I think we agree on the key issue for now. But other notes:

> Simplifying existing passes: It would be good to take a look at simplifying DCE, MergeBlocks, and any pass that expects flat IR by having them operate on Poppy IR, either having them run stackification and unstackification as subpasses or depending on the pass runner to insert conversions. 

I'm skeptical about this (I wrote a similar comment [here](https://github.com/WebAssembly/binaryen/issues/3090#issuecomment-686603025) just now actually, but this expands on it). Converting IR back and forth would have a cost in speed. Binaryen's optimization pipeline is designed to be minimal and fast, with most optimizations just removing or altering code (i.e. no allocations and no major restructurings which could affect locality). Basic optimizations like DCE and MergeBlocks should have that behavior, as well as every pass we run in `-O1` (higher levels do some much heavier things, of course).

Note also that I'm unsure how much Poppy would help with them. They do more than natively remove dead code or naively merge blocks - they also do a bunch of things related to those which are easier and more efficient to do at the same time.

Reply:
Very reasonable. Even if the code could be much simpler if it used Poppy IR (and I am curious about that), I agree that the current complexity is probably worth saving the overhead of performing the translations to and from Poppy IR.